controllers/cloud.redhat.com/hashcache/hashcache.go

```diff
@@ -0,0 +1,234 @@
+package hashcache
+
+import (
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"sort"
+	"sync"
+
+	crd "github.com/RedHatInsights/clowder/apis/cloud.redhat.com/v1alpha1"
+	"github.com/RedHatInsights/clowder/controllers/cloud.redhat.com/errors"
+	"github.com/RedHatInsights/clowder/controllers/cloud.redhat.com/object"
+	core "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func generateHashFromData(data []byte) (hash string) {
+	h := sha256.New()
+	h.Write([]byte(data))
+	hash = fmt.Sprintf("%x", h.Sum(nil))
+	return
+}
+
+type Ident struct {
+	NN   types.NamespacedName
+	Type string
+}
+
+type HashObject struct {
+	Hash      string
+	ClowdApps map[types.NamespacedName]bool
+	ClowdEnvs map[types.NamespacedName]bool
+}
+
+type HashCache struct {
+	data map[Ident]*HashObject
+	lock sync.RWMutex
+}
+
+func NewHashCache() HashCache {
+	return HashCache{
+		data: map[Ident]*HashObject{},
+		lock: sync.RWMutex{},
+	}
+}
+
+func NewHashObject(hash string) HashObject {
+	return HashObject{
+		Hash:      hash,
+		ClowdApps: map[types.NamespacedName]bool{},
+		ClowdEnvs: map[types.NamespacedName]bool{},
+	}
+}
+
+type ItemNotFoundError struct {
+	item string
+}
+
+func (a ItemNotFoundError) Error() string {
+	return fmt.Sprintf("item [%s] not found", a.item)
+}
+
+func (hc *HashCache) Read(obj client.Object) (*HashObject, error) {
+	var oType string
+	switch obj.(type) {
+	case *core.ConfigMap:
+		oType = "ConfigMap"
+	case *core.Secret:
+		oType = "Secret"
+	}
+
+	id := Ident{NN: types.NamespacedName{Name: obj.GetName(), Namespace: obj.GetNamespace()}, Type: oType}
+	hc.lock.RLock()
+	defer hc.lock.RUnlock()
+	v, ok := hc.data[id]
+	if !ok {
+		return nil, ItemNotFoundError{item: fmt.Sprintf("%s/%s", id.NN.Name, id.NN.Namespace)}
+	}
+	return v, nil
+}
+
+func (hc *HashCache) RemoveClowdObjectFromObjects(obj client.Object) {
+	hc.lock.Lock()
+	defer hc.lock.Unlock()
+
+	for _, v := range hc.data {
+		switch obj.(type) {
+		case *crd.ClowdEnvironment:
+			delete(v.ClowdEnvs, types.NamespacedName{
+				Name:      obj.GetName(),
+				Namespace: obj.GetNamespace(),
+			})
+		case *crd.ClowdApp:
+			delete(v.ClowdApps, types.NamespacedName{
+				Name:      obj.GetName(),
+				Namespace: obj.GetNamespace(),
+			})
+		}
+	}
+}
+
+func (hc *HashCache) CreateOrUpdateObject(obj client.Object) (bool, error) {
+	// NEED TO PASS IN Q HERE TO GET IT TO ALSO TRIGGER
```

Comment:
Is this comment still operative? I don't know what it means but it is in all caps which makes it look very important. If I encountered this while working I wouldn't know what it meant.